contracts/BalanceTrackerFixedPriceBase.sol

```diff
@@ -0,0 +1,242 @@
+
+        // Check the request delivery rate for a fixed price
+        if (balance < maxDeliveryRate) {
+            // Get balance difference
+            uint256 balanceDiff = maxDeliveryRate - balance;
+            // Adjust balance
+            balance += _getRequiredFunds(requester, balanceDiff);
+        }
+
+        // Adjust account balance
+        balance -= maxDeliveryRate;
+        mapRequesterBalances[requester] = balance;
+    }
+
+    /// @dev Finalizes mech delivery rate based on requested and actual ones.
+    /// @param mech Delivery mech address.
+    /// @param requester Requester address.
+    /// @param requestId Request Id.
+    /// @param maxDeliveryRate Requested max delivery rate.
+    function finalizeDeliveryRate(address mech, address requester, uint256 requestId, uint256 maxDeliveryRate) external {
+        // Check for marketplace access
+        if (msg.sender != mechMarketplace) {
+            revert MarketplaceOnly(msg.sender, mechMarketplace);
+        }
+
+        // Get actual delivery rate
+        uint256 actualDeliveryRate = IMech(mech).getFinalizedDeliveryRate(requestId);
+
+        // Check for zero value
+        if (actualDeliveryRate == 0) {
+            revert ZeroValue();
+        }
+
+        uint256 rateDiff;
+        if (maxDeliveryRate > actualDeliveryRate) {
+            // Return back requester overpayment debit
+            rateDiff = maxDeliveryRate - actualDeliveryRate;
+            mapRequesterBalances[requester] += rateDiff;
+        } else {
+            actualDeliveryRate = maxDeliveryRate;
+        }
+
+        // Record payment into mech balance
+        mapMechBalances[mech] += actualDeliveryRate;
+
+        emit MechPaymentCalculated(mech, requestId, actualDeliveryRate, rateDiff);
+    }
+
+    function _drain(uint256 amount) internal virtual;
+
+    /// @dev Drains collected fees by sending them to a Buy back burner contract.
+    function drain() external {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        uint256 localCollectedFees = collectedFees;
+
+        // Check for zero value
+        if (localCollectedFees == 0) {
+            revert ZeroValue();
+        }
+
+        collectedFees = 0;
+
+        // Drain
+        _drain(localCollectedFees);
+
+        _locked = 1;
+    }
+
+    function _withdraw(uint256 balance) internal virtual;
+
+    /// @dev Processes mech payment by withdrawing funds.
+    function processPayment() external returns (uint256 mechPayment, uint256 marketplaceFee) {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Get mech balance
+        uint256 balance = mapMechBalances[msg.sender];
+        // TODO minimal balance value to account for the round-off
+        if (balance == 0 || balance < 10_000) {
+            revert InsufficientBalance(balance, 10_000);
+        }
```

Comment:
Need to define minimal value here to compare against.

Reply:
The rest is unchanged.

Reply:
My idea was to do the rounding up (ceil, not floor). This way, we don't need to define a hardcode on the balance as the marketplaceFee is always strictly positive (e.g. 1% fee, 1 cent withdrawn -> fee is 1 cent; 50 cent withdrawn -> fee is 1 cent, 1 dollar withdrawn -> fee is 1 cent